[v4.1] Added WebSocket subsection #649

Racater opened the pull request:
<!---
IMPORTANT NOTES:
- Changes should always be made only in the raw .md files and not in the CSV, JSON, XLSX, PDF, DOCX files, etc.
- Please do not open a pull request without first opening an associated issue.
- Please carry out all discussion in the associated issue only.
-->

This Pull Request relates to issue #649

Racater commented:
Please review the changes, As per the issue, this should be added only to 4.1, not to 4.0.2. Feel free to suggest improvements or ask for clarifications.

Sjord commented:
> Ensure that

Most rules start with "Verify that". Is that better?

> | **13.5.3** | verify that

"Verify" should be with a capital.

> Verify that WebSocket messages are properly sanitized and encoded.

What does this mean exactly? Who should do the sanitation? What values are dangerous in a websocket message?

> Ensure rate limiting is in place for WebSocket messages

Is this required? I don't think we currently require this for HTTP requests. Why would it be more important for websocket requests?

> If you are using session or channel tokens in webSockets, ensure that the tokens are obtained from a response to secure POST request only, and are not sent by the server through the Websocket connection.

What is this about? Why is sending tokens in a websocket insecure?

> If you are using session or channel tokens in webSockets, ensure that the tokens possess at least 64 bits of entropy and are generated using approved cryptographic algorithms.

Does this need a websocket-specific rule, or is 3.2.2 sufficient?

Racater commented:
@Sjord thanks a lot for the review.

> Most rules start with "Verify that". Is that better?
I have changed all instances of "ensure" to "verify" and corrected the capitalization for consistency with the rest of the document.

> Verify that WebSocket messages are properly sanitized and encoded
This rule comes directly from the OWASP website recommendations: https://owasp.org/www-project-web-security-testing-guide/v41/4-Web_Application_Security_Testing/11-Client_Side_Testing/10-Testing_WebSockets . As websocket messages are an unsafe data input, it should of course be sanitized. Specific sanitation steps are application specific. The data sanitation must be server side.

> Ensure rate limiting is in place for WebSocket messages
Websockets messages are often less well handled by reverse proxies, such as NGINX. Indeed the websocket data is in tcp stream, not in transaction. Websocket data can probably not be separated by packet or transaction and can not be filtered this way. This makes DoS easier on websocket. However, since there is no mention of rate limiting for http, it makes sense to remove this item.

> If you are using session or channel tokens in webSockets, ensure that the tokens are obtained from a response to secure POST request only, and are not sent by the server through the Websocket connection.
The problem with the server sending the session token via websocket is that it bypasses the CORS protection
offered by the browser. Moreover, since starting a session (getting a session token) is a state-changing operation on the server side, a POST request should be used. See http://www.adambarth.com/papers/2008/barth-jackson-mitchell-b.pdf .

> If you are using session or channel tokens in webSockets, ensure that the tokens possess at least 64 bits of entropy and are generated using approved cryptographic algorithms.

This has been added as some applications use a different tokens for webSockets and for http. It is good that users check websocket tokens entropy and not just 3.2.2.

Thanks again for the review. Like you suggested, maybe some elements could be removed if deemed too specific or redundant.

tghosth commented on 4.0/en/0x21-V13-API.md:
```diff
@@ -56,6 +56,18 @@ Note: Due to issues with XXE attacks against DTDs, DTD validation should not be
+
+| # | Description | L1 | L2 | L3 | CWE |
+| :---: | :--- | :---: | :---:| :---: | :---: |
+| **13.5.1** | Verify that only authenticated users can open WebSocket connections. | ✓ | ✓ | ✓ | 306 |
```
Will web sockets only ever be for authenticated uses? Could they not also be used for unauthenticated purposes as well?

Racater replied:
Good point. Even though [some resources](https://www.freecodecamp.org/news/how-to-secure-your-websocket-connections-d0be0996c556/#-4-authenticate-users-before-ws-connection-establishes) say that "establishing a websocket connection is a bit expensive anyway", ASVS should in my opinion also allow unauthenticated websockets. I modified the rule.

Note that as stated in [this paper](https://www.researchgate.net/publication/315858773_Slowloris_DoS_Countermeasure_over_WebSocket), websocket is vulnerable to slowloris (not closing connections). It might be worth it to add another rule for connection limiting, to investigate...

tghosth commented on 4.0/en/0x21-V13-API.md:
```diff
@@ -56,6 +56,18 @@ Note: Due to issues with XXE attacks against DTDs, DTD validation should not be
+| :---: | :--- | :---: | :---:| :---: | :---: |
+| **13.5.1** | Verify that only authenticated users can open WebSocket connections. | ✓ | ✓ | ✓ | 306 |
+| **13.5.2** | Verify that WebSocket Secure (wss) is used for all WebSocket connections. | ✓ | ✓ | ✓ | 319 |
+| **13.5.3** | Verify that the Origin header is checked in the initial HTTP WebSocket handshake. | ✓ | ✓ | ✓ | 346 |
```
Checked for what?

Racater replied:
clarified

tghosth commented on 4.0/en/0x21-V13-API.md:
```diff
@@ -56,6 +56,18 @@ Note: Due to issues with XXE attacks against DTDs, DTD validation should not be
+| **13.5.1** | Verify that only authenticated users can open WebSocket connections. | ✓ | ✓ | ✓ | 306 |
+| **13.5.2** | Verify that WebSocket Secure (wss) is used for all WebSocket connections. | ✓ | ✓ | ✓ | 319 |
+| **13.5.3** | Verify that the Origin header is checked in the initial HTTP WebSocket handshake. | ✓ | ✓ | ✓ | 346 |
+| **13.5.4** | Verify that WebSocket messages are properly sanitized and encoded server-side. | ✓ | ✓ | ✓ | 116 |
```
Sanitization and/or encoding should be performed before input is used in a different context, e.g. incorporated into a file path, used in a SQL query, written back to a web page's HTML, etc. It is not clear what the purpose is here.

Do we mean input validation to ensure that messages received conform to expected formats?

Racater replied:
You are right. What do you think of such a formulation:

> Verify that websocket input from the client is sanitized server-side before being processed

This item is in part a repetition of section 5 of ASVS. The aim is to make sure that Websocket inputs are taken into account, as they are not mentioned in section 5. Maybe it's better to modify section 5 though?

Racater replied:
I removed the rule altogether, to avoid duplication.

tghosth commented on 4.0/en/0x21-V13-API.md:
```diff
@@ -56,6 +56,18 @@ Note: Due to issues with XXE attacks against DTDs, DTD validation should not be
+| **13.5.4** | Verify that WebSocket messages are properly sanitized and encoded server-side. | ✓ | ✓ | ✓ | 116 |
+| **13.5.5** | Verify that rate limiting is in place for WebSocket messages | ✓ | ✓ | ✓ | 770 |
+| **13.5.6** | If you are using session or channel tokens in webSockets, verify that the tokens are obtained from a response to secure POST request only, and are not sent by the server through the Websocket connection. | ✓ | ✓ | ✓ | 319 |
+| **13.5.7** | If you are using session or channel tokens in webSockets, verify that the tokens possess at least 64 bits of entropy and are generated using approved cryptographic algorithms. | ✓ | ✓ | ✓ | 331 |
```
Would it be better to reference the session token guidance contained elsewhere in this standard? Are these requirements different to other session token requirements which we already have?

Racater replied:
You are right. This part is a repetition of section 3.2 of the standard. However, websocket session tokens are often different than cookie session tokens. They should be **checked separately**.

What do you think would be the best solution?

- Reference section 3.2 in the 13.5.7 requirement?
    - Such as done in rule 3.2.3. It makes the requirement explicit, but still creates a bit of repetition
- Replace 13.5.7 by a sentence in the 13.5 section text?
    - It removes the repetition, but creates a "hidden" requirement in the paragraph text. As such we are not sure that this specific token will be checked.

tghosth commented on 4.0/en/0x21-V13-API.md:
```diff
@@ -56,6 +56,18 @@ Note: Due to issues with XXE attacks against DTDs, DTD validation should not be
+| **13.5.3** | Verify that the Origin header is checked in the initial HTTP WebSocket handshake. | ✓ | ✓ | ✓ | 346 |
+| **13.5.4** | Verify that WebSocket messages are properly sanitized and encoded server-side. | ✓ | ✓ | ✓ | 116 |
+| **13.5.5** | Verify that rate limiting is in place for WebSocket messages | ✓ | ✓ | ✓ | 770 |
+| **13.5.6** | If you are using session or channel tokens in webSockets, verify that the tokens are obtained from a response to secure POST request only, and are not sent by the server through the Websocket connection. | ✓ | ✓ | ✓ | 319 |
```
Are you talking about initially receiving the token here, i.e. how it is produced in the first place? Please can you clarify here.

Racater replied:
clarified

tghosth reviewed (changes requested):
Thanks for your efforts on this. Please can you review and action my comments.

Racater commented:
thank you for your review @tghosth . I will work on it and commit updates

elarlang commented:
> V13.5.4| Verify that WebSocket messages are properly sanitized and encoded server-side.

Isn't it covered by 5.2.1, 5.3.1, 5.3.3? If not, what is the difference?

> V13.5.5 Verify that rate limiting is in place for WebSocket messages

Isn't it covered by 11.1.4? If not, what is the difference?

Racater commented:
@tghosth I updated the PR based on your comments. I also reordered the rules in a more logical order. 

Note that there is still repetitions as follows:
- Rule 13.5.3 is basically section 5 (Validation) applied to Websockets
- Rule 13.5.6 is rules from 3.2 (tokens) applied to Websockets

I am still a bit unsure how to handle the repetition and any suggestion would be welcome.

elarlang commented:
There are not clear rules (based on my knowledge) and there are some duplicate requirements (in files section for example) because of that.

I watch WebSocket as communication channel - it does not have anything to do how data is used later, I think we don't need separate requirement for "data from websockets must be validated / sanitized / converted also". It must be done anyway.

With anti-automation requirement it's not that clear, but in my opinion proposed requirement should be part of current 11.1.4.

> Rule 13.5.6 is rules from 3.2 (tokens) applied to Websockets

What is 3.2 here? Current subcategory 3.2 is "V3.2 Session Binding Requirements" and it does not make sense in this context.

Racater commented:
Good point, I removed the sanitation rule and added more details about WebSocket-specific tokens.

Racater commented:
done, waiting for your input

jmanico commented:
Is this ready to be merged?

elarlang commented:
Some feedback/comments on style/language:
* "whitelist" may be alarming nowadays, something like "allow list" or "allowed domains" instead
* Probably every requirement should start with "Verify that" like all other requirements in ASVS
* "you are using" -> "being used", "is in use"

jmanico commented:
I am fixing the language now - standby.

jmanico commented:
Language fixed per dbfd721
